ZeRO blog post (#71)
huggingface/blog: zero-deepspeed-fairscale.md

---
title: "Fit More and Train Faster With ZeRO via DeepSpeed and FairScale"
thumbnail: /blog/assets/09_zero_deepspeed_fairscale/thumbnail.png
---

<h1 class="no-top-margin">Porting fairseq wmt19 translation system to transformers</h1>

<div class="blog-metadata">

>     <small>Published November 3, 2020.</small>

Reviewer: Should this be updated manually when this is merged?
```suggestion
    <small>Published November 3, 2020.</small>
```

    <a target="_blank" class="btn-readme" href="https://github.com/huggingface/blog/blob/master/zero-deepspeed-fairscale.md">
        <img src="/front/assets/icon-github.svg">
        Update on GitHub
    </a>
</div>

<div class="author-card">
    <a href="/stas">
        <img class="avatar avatar-user" src="/blog/assets/09_zero_deepspeed_fairscale/stas-bekman-300x300.jpg">
        <div class="bfc">
            <code>stas</code>
            <span class="fullname">Stas Bekman</span>
            <span class="mono-label" style="color: var(--text-main);">guest</span>
        </div>
    </a>
</div>

> ##### A guest blog post by Stas Bekman

Reviewer 2: ```suggestion
##### A guest blog post by Hugging Face fellow Stas Bekman
```
Not sure if you want to advertise your unofficial title here, feel free to ignore!

# **Fit More and Train Faster With ZeRO via DeepSpeed and FairScale**

> As of recent Machine Learning models have been growing much faster than the amount of GPU memory added to newly released cards and many users are unable to train or even just load some of those huge models onto their hardware. While there is an ongoing effort to distill some of those huge models to be of a more manageable size -- that effort isn't producing models small enough soon enough.

Reviewer 2: ```suggestion
As recent Machine Learning models have been growing much faster than the amount of GPU memory added to newly released cards, many users are unable to train or even just load some of those huge models onto their hardware. While there is an ongoing effort to distill some of those huge models to be of a more manageable size -- that effort isn't producing models small enough soon enough.
```

In the fall of 2019 Samyam Rajbhandari, Jeff Rasley, Olatunji Ruwase and Yuxiong He published a paper:

> [ZeRO: Memory Optimizations Toward Training Trillion Parameter Models](https://arxiv.org/abs/1910.02054), which contains a plethora of ingenious new ideas on how one could make their hardware do much more than what it was thought possible before. A short time later [DeepSpeed](https://github.com/microsoft/deepspeed) has been released that gave to the world the open source implementation of most of the ideas in that paper (a few ideas are still in works) and in parallel the team from Facebook released [FairScale](https://github.com/facebookresearch/fairscale/) which also implemented some of the core ideas from the ZeRO paper.

Reviewer 2: ```suggestion
[ZeRO: Memory Optimizations Toward Training Trillion Parameter Models](https://arxiv.org/abs/1910.02054), which contains a plethora of ingenious new ideas on how one could make their hardware do much more than what it was thought possible before. A short time later [DeepSpeed](https://github.com/microsoft/deepspeed) has been released that gave to the world the open source implementation of most of the ideas in that paper (a few ideas are still in works) and in parallel a team from Facebook released [FairScale](https://github.com/facebookresearch/fairscale/) which also implemented some of the core ideas from the ZeRO paper.
```

> If you use the HF Trainer, as of `transformers` v4.2.0 you have the initial experimental support for DeepSpeed's and FairScale's ZeRO features. The new `--sharded_ddp` and `--deepspeed` command line trainer arguments provide FairScale and DeepSpeed integration respectively. Here is [the full documentation](https://huggingface.co/transformers/master/main_classes/trainer.html#trainer-integrations).

Reviewer 2: ```suggestion
If you use the Hugging Face Trainer, as of `transformers` v4.2.0 you have the initial experimental support for DeepSpeed's and FairScale's ZeRO features. The new `--sharded_ddp` and `--deepspeed` command line `Trainer` arguments provide FairScale and DeepSpeed integration respectively. Here is [the full documentation](https://huggingface.co/transformers/master/main_classes/trainer.html#trainer-integrations).
```

Reviewer 2: Also "initial experimental" is a bit too much? Maybe just one of those adjectives is enough?

This blog post will describe how you can benefit from ZeRO regardless of whether you own just a single GPU or a whole stack of them.

> # Huge Speedups with multi-gpu setups

Reviewer 2: ```suggestion
# Huge Speedups with multi-GPU setups
```

> Let's do a small finetuning with translation task experiment, using `t5-large` model and  `finetune_trainer.py` script which you can find under [`examples/seq2seq`](https://github.com/huggingface/transformers/tree/master/examples/seq2seq) in the `transformers` github repo.

Reviewer 2: ```suggestion
Let's do a small finetuning with translation task experiment, using a `t5-large` model and the `finetune_trainer.py` script which you can find under [`examples/seq2seq`](https://github.com/huggingface/transformers/tree/master/examples/seq2seq) in the `transformers` GitHub repo.
```

We have 2x 24GB (Titan RTX) GPUs to test with.

> This is just a proof of concept benchmarks so surely things can be improved further, so we will benchmark on a small sample of train=2000, eval=500 items to perform the comparisons.  Evaluation does by default beam search size=4, so it's slower than training with the same number of samples, that's why 4x less eval items were used in these tests.

Reviewer 2: Mmm, either use code formatting with the full args names: `n_train=2000`, `n_eval=500`, `beam_search_size=4` or just use plain English.

Here are the key command line arguments of our baseline:
```
export BS=16
python -m torch.distributed.launch --nproc_per_node=2 ./finetune_trainer.py \
--model_name_or_path t5-large --n_train 2000 --n_val 500 \
--per_device_eval_batch_size $BS --per_device_train_batch_size $BS \
--task translation_en_to_ro [...]
```

We are just using the DistributedDataParallel (DDP) and nothing else to boost the performance for the baseline. I was able to fit a batch size (BS) of 16 before hitting Out of Memory (OOM) error.

> Note, that for simplicity and to make it easier to understand, I have shown only the important for this demonstration command line arguments. You will find the complete command line at

Reviewer 3: ```suggestion
Note, that for simplicity and to make it easier to understand, I have only shown the important command line arguments for this demonstration. You will find the complete command line at
```

Author: It doesn't sound good. Did you mean to write it as:

> I have only shown the command line arguments important for this demonstration.

or perhaps:

> I have only shown the command line arguments relevant to this demonstration.

perhaps?

> ```

Reviewer: This is a bit hard to parse

Author: Agreed. I rewrite it to remove almost all cl args, leaving the important ones and linked to the full command line instead.

> We are just using the DistributedDataParallel (DDP) and nothing else to boost the performance for the baseline. I was able to fit a BatchSize (BS) of 16 before hitting OOM.

Reviewer 2: ```suggestion
We are just using the DistributedDataParallel (DDP) and nothing else to boost the performance for the baseline. I was able to fit a batch size (BS) of 16 before hitting out of memory (OOM).
```

Author: yes and needed to add 'error' if that' the approach.

[post](https://github.com/huggingface/transformers/issues/8771#issuecomment-759248400).

Next we are going to re-run the benchmark every time adding one of the following:

1. `--fp16`
2. `--sharded_ddp` (fairscale)
3. `--sharded_ddp --fp16` (fairscale)
4. `--deepspeed` without cpu offloading
5. `--deepspeed` with cpu offloading

Since the key optimization here is that each technique deploys GPU RAM more efficiently, we will try to continually increase the batch size and expect the training and evaluation to complete faster (while keeping the metrics steady or even improving some, but we won't focus on these here).

> Remember that training and evaluation stages are very different from each other, because during training models weights are being modified, gradients are being calculated, and optimizer states are stored. During evaluation, none of these happen, but in this particular task of translation the model will try to search for the best hypothesis, so it actually has to do multiple runs before it's satisfied. That's why it's not fast, especially when a model is large.

Reviewer 3: ```suggestion
Remember that training and evaluation stages are very different from each other, because during training model weights are being modified, gradients are being calculated, and optimizer states are stored. During evaluation, none of these happen, but in this particular task of translation the model will try to search for the best hypothesis, so it actually has to do multiple runs before it's satisfied. That's why it's not fast, especially when a model is large.
```

> Remember that training and evaluation stages are very different from each other, because during training models weights are being modified, gradients are being calculated, and optimizer states are stored. During evaluation, none of these happen, but in this particular task of translation the model will try to search for the best hypothesis, so it actually has to do multiple runs before it's satisfied, so it's not fast, especially when a model is large.

Reviewer 2: ```suggestion
Remember that training and evaluation stages are very different from each other, because during training models weights are being modified, gradients are being calculated, and optimizer states are stored. During evaluation, none of these happen, but in this particular task of translation the model will try to search for the best hypothesis, so it actually has to do multiple runs before it's satisfied. That's why it's not fast, especially when a model is large.
```

> Let's look at the results of these 6 test runs:

Reviewer 2: ```suggestion
Let's look at the results of these six test runs:
```

> | Method                    | max BS |  train time |   eval time |
> |---------------------------+--------+-------------+-------------|
> | baseline                  |     16 |     30.9458 |     56.3310 |
> | fp16                      |     20 |     21.4943 |     53.4675 |
> | sharded_ddp               |     30 |     25.9085 |     47.5589 |
> | sharded_ddp+fp16          |     30 |     17.3838 |     45.6593 |
> | deepspeed w/o cpu offload |     40 | **10.4007** |     34.9289 |
> | deepspeed w/ cpu offload  | **50** |     20.9706 | **32.1409** |

Reviewer: This doesn't render well on GitHub nor does it in my IDE; I had to replace the "+" by "|" in the header:
```suggestion
| Method                    | max BS |  train time |   eval time |
|---------------------------|--------|-------------|-------------|
| baseline                  |     16 |     30.9458 |     56.3310 |
| fp16                      |     20 |     21.4943 |     53.4675 |
| sharded_ddp               |     30 |     25.9085 |     47.5589 |
| sharded_ddp+fp16          |     30 |     17.3838 |     45.6593 |
| deepspeed w/o cpu offload |     40 | **10.4007** |     34.9289 |
| deepspeed w/ cpu offload  | **50** |     20.9706 | **32.1409** |
```

> It's easy to see that both FairScale and DeepSpeed provide great improvements over the baseline, in the total train/eval time but also in the batch size. DeepSpeed implements more magic as of this writing and seems to be the short term winner, but Fairscale is easier to deploy. For DeepSpeed you need to write a simple configuration file and change your command line's launcher, with Fairscale you only need to add the `--sharded_ddp` command line argument, so you may want to try it first as it's the most low-hanging fruit.

Reviewer 2: ```suggestion
It's easy to see that both FairScale and DeepSpeed provide great improvements over the baseline, in the total train and evaluation time but also in the batch size. DeepSpeed implements more magic as of this writing and seems to be the short term winner, but Fairscale is easier to deploy. For DeepSpeed you need to write a simple configuration file and change your command line's launcher, with Fairscale you only need to add the `--sharded_ddp` command line argument, so you may want to try it first as it's the most low-hanging fruit.
```

> Following the 80:20 rule, I have only spent a few hours on these benchmarks and I haven't tried to squeeze every MB and second by refining the command line arguments and configuration, since it's pretty obvious from the simple table what you'd want to try next. When you will face a real project that will be running for hours and perhaps days, definitely spend more time to make sure you use the most optimal possible hyper parameters to get your job done faster and at a minimal cost.

Reviewer 2: ```suggestion
Following the 80:20 rule, I have only spent a few hours on these benchmarks and I haven't tried to squeeze every MB and second by refining the command line arguments and configuration, since it's pretty obvious from the simple table what you'd want to try next. When you will face a real project that will be running for hours and perhaps days, definitely spend more time to make sure you use the most optimal possible hyper-parameters to get your job done faster and at a minimal cost.
```

If you would like to experiment with this benchmark yourself or want to know more details about the hardware and software used to run it, please, refer to this [post](https://github.com/huggingface/transformers/issues/8771#issuecomment-759248400).

> # Fitting A Huge Model Onto 1 GPU

Reviewer 2: ```suggestion
# Fitting A Huge Model Onto one GPU
```
You didn't cap the other titles by the way.

Author: I went over and made sure they all are consistent

While Fairscale gives us a boost only with multiple GPUs, DeepSpeed has a gift even for those of us with a single GPU.

Let's try the impossible - let's train [t5-3b](https://huggingface.co/t5-3b) on a 24GB RTX-3090 card.

First let's try to finetune the huge `t5-3b` using the normal single GPU setup:
```
export BS=1; rm -r output_dir; CUDA_VISIBLE_DEVICES=0 PYTHONPATH=../../src USE_TF=0 \
./finetune_trainer.py --model_name_or_path t5-3b --output_dir output_dir --adam_eps 1e-06 \
--data_dir wmt_en_ro --do_eval --do_train --evaluation_strategy=steps --freeze_embeds \
--label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step --logging_steps 1000 \
--max_source_length 128 --max_target_length 128 --num_train_epochs 1 --overwrite_output_dir \
--per_device_eval_batch_size $BS --per_device_train_batch_size $BS --predict_with_generate \
--eval_steps 25000  --sortish_sampler --task translation_en_to_ro  \

> --val_max_target_length 128 --warmup_steps 5 --n_train 60 --n_val 10 --fp16

Reviewer 2: Same comment as above on the command.

```
No cookie, even with BS=1 we get:
```
RuntimeError: CUDA out of memory. Tried to allocate 64.00 MiB (GPU 0; 23.70 GiB total capacity;
21.37 GiB already allocated; 45.69 MiB free; 22.05 GiB reserved in total by PyTorch)
```

> Now update your `transformers` to v4.0.2 or higher, then install DeepSpeed:

Reviewer: ```suggestion
Now update your `transformers` to v4.2.0 or higher, then install DeepSpeed:
```

```
pip install deepspeed
```

and let's try again:
```
export BS=20; rm -r output_dir; CUDA_VISIBLE_DEVICES=0 PYTHONPATH=../../src USE_TF=0 deepspeed \
--num_gpus=1 ./finetune_trainer.py --model_name_or_path t5-3b --output_dir output_dir \
--adam_eps 1e-06 --data_dir wmt_en_ro --do_eval --do_train --evaluation_strategy=steps \
--freeze_embeds --label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step \
--logging_steps 1000 --max_source_length 128 --max_target_length 128 --num_train_epochs 1 \
--overwrite_output_dir --per_device_eval_batch_size $BS --per_device_train_batch_size $BS \
--predict_with_generate --eval_steps 25000  --sortish_sampler --task translation_en_to_ro \
--val_max_target_length 128 --warmup_steps 5 --n_train 60 --n_val 10 \

> --deepspeed ds_config_1gpu.json --fp16

Reviewer: Here too, it's hard to understand what you changed in this command from the previous one as it just looks like a wall of text on GitHub

Here are the relevant results:
```
2021-01-12 19:06:31 | INFO | __main__ |   train_n_objs = 60
2021-01-12 19:06:31 | INFO | __main__ |   train_runtime = 8.8511
2021-01-12 19:06:35 | INFO | __main__ |   val_n_objs = 10
2021-01-12 19:06:35 | INFO | __main__ |   val_runtime = 3.5329
```
We can't compare these to the baseline, since the baseline won't even start and immediately fail with OOM.

Simply amazing!

I did only a tiny sample since I was primarily interested in being able to train and evaluate with this huge model that normally won't fit onto a 24GB GPU.

If you would like to experiment with this benchmark yourself or want to know more details about the hardware and software used to run it, please, refer to this [post](https://github.com/huggingface/transformers/issues/8771#issuecomment-759176685).

# The Magic Behind ZeRO

Since `transformers` only integrated these fabulous solutions and wasn't part of their invention I will share the resources where you can discover all the details for yourself. But here are a few quick insights that may help understand how ZeRO manages these amazing feats.

> The key feature is 3D parallelism, which combines data parallelism (DP), pipeline parallelism (PP), and tensor-slicing model parallelism (MP). Here is an illustration from this [blog post](https://www.microsoft.com/en-us/research/blog/deepspeed-extreme-scale-model-training-for-everyone/) that attempts to visualize this very complex approach:

Reviewer 4: @stas00 Really cool blog post. Want to suggest some correction here. 3D parallelism is not a feature of ZeRO.  The key feature of ZeRO is the partitioning of parameter, gradient and optimizer state in data parallel training, which saves memory over the standard data parallel training which replicates these model states. The Zero Redundancy refers to the removal of memory redundancy in data parallel training

In terms of relation between 3D parallelism and ZeRO, ZeRO can be used with 3D parallelism, where we can use ZeRO-powered data parallelism as one of the parallel dimensions of 3D parallelism. If you feel its helpful to have a chat, I am would be happy to do so :)

Author: Well, the whole concept of model parallelism is very can-mean-anything, and your way of partitioning is a kind of MP just done as a horizontal split.

I will edit that part for sure - thank you for flagging that!

And yes, I'd love to discuss this with you and improve the post - please let me know how the best for us to connect - slack, or stas@stason.org by email, or any other way that works the best for you. Thank you so much!

Author: BTW, quoting your own blog: https://www.microsoft.com/en-us/research/blog/deepspeed-extreme-scale-model-training-for-everyone/
> Trillion parameter model training with 3D parallelism: DeepSpeed enables a flexible combination of three parallelism approaches—ZeRO-powered data parallelism, pipeline parallelism, and tensor-slicing model parallelism. 3D parallelism adapts to the varying needs of workload requirements to power extremely large models with over a trillion parameters while achieving near-perfect memory-scaling and throughput-scaling efficiency. In addition, its improved communication efficiency allows users to train multi-billion-parameter models 2–7x faster on regular clusters with limited network bandwidth.

and then later:

> DeepSpeed has combined three powerful technologies to enable training trillion-scale models and to scale to thousands of GPUs: data parallel training, model parallel training, and pipeline parallel training.

Right?

Unless what you're trying to say that, yes, there is 3D parallelism, but it's not the key feature. And the key innovation is tensor-slicing model parallelism via gradient and optimizer state sharding and parameter sharding coming soon.

Reviewer 5: Maybe let me describe each feature in the parallelism space separately to hopefully clear up some confusion. 

1) ZeRO is the technique we developed to partition optimizer, gradient, and/or parameter states across data parallel ranks. 
2) Pipeline parallelism partitions layers of a model into stages that can then be processed in parallel. It can be used along side traditional data parallelism or can be also be combined with ZeRO.
3) Tensor slicing/op sharding as done in Megatron-LM allows for specific ops in a model to be split across multiple ranks, thus reducing the memory footprint of specific components of a model (e.g., [RowParallelLinear](https://github.com/microsoft/DeepSpeedExamples/blob/master/Megatron-LM/mpu/layers.py#L252)). 
4) 3D parallelism is the concept we introduced in the press release you quote above. This concept is the combination of all 3 of the above types of parallelism.

I believe the focus on the integration so far has been primarily in enabling number 1 from above, correct?

Reviewer 4: Adding to @jeffra, one way I find it helpful to think about ZeRO is simply as data parallel training but with distributed data storage. The computation on each GPU is exactly the same as data parallel training, but the parameter, gradients and optimizer states are stored in a distributed/partition fashion across all the GPUs and fetched only when needed. It doesn't really matter if the partition is horizontal or vertical. This is quiet different from model parallel techniques such as tensor slicing or pipeline parallelism which divide not just the parameter, gradients and optimizer states, but also the computation vertically and horizontally, respectively.

Author: Thank you for these clarifications, @jeffra and @samyam!

1. Yes, we have Zero stage 1+2 and some of its other parts integrated (e.g. cpu offload). So we have ZeRO - check!

2. So you're saying that PP isn't being used by default? So `train_micro_batch_size_per_gpu` is only relevant if we explicitly implement use PP, right? which I don't think can be used with ZeRO, or am I wrong?

3. Tensor slicing/op sharding - I guess this is where I'm most confused - isn't param/optim/gradient sharding the same as tensor slicing/op sharding? In both cases you partition tensors and spread them out to different GPUs - so there must be a nuance that I'm missing here.
 
@samyam, your wording is great! I will rewrite that section using your explanation. I guess I will use this diagram then https://user-images.githubusercontent.com/10676103/104272403-df472d00-5451-11eb-94da-53017aa6631d.png
but it's trickier to explain.

Author: I think DeepSpeed would benefit a lot from having a sort of a visual map showing how the different components fit together. 
* Which can inter-operate, which cannot. 
* Which are part of ZeRO, which are stand-apart.
* Which parts stack to increase speed, which to reduce memory footprint

The blog posts have a lot of diagrams, but they only convey partial picture.

Author: OK, I switched to a different diagram and rewrote that section based on your clarifications. Please kindly have a look.

If you feel anything else would benefit the user please don't hesitate to suggest.

thank you!

Author: Please let's continue this discussion at https://github.com/microsoft/DeepSpeed/issues/673 so that it'd be easier to find your super useful commentary in the future (github hides comments once things are resolved). Thank you!

Author: @samyam and/or @jeffra - would it be possible to validate that you're happy with the current wording? I think the blog is ready to go online otherwise. Thank you very much for this contribution!

Author: @samyam, I'd love for one of you to address this issue: https://github.com/microsoft/DeepSpeed/issues/674

I'm being asked about it and I can't give any good answers. I know you said it should work with vertical MP, but it doesn't work out of box and I'd love for you to tell me what needs to be done to make it work.

We have vertical MP implemented in t5/gpt/bart so it'd be easy for me to test.

![3D parallelism](./assets/09_zero_deepspeed_fairscale/3d-parallelism.png]

Let's briefly touch on each of the 3 types:

> 1. **Data Parallelism** is what you get from `DataParallel` - we replicate the model to each GPU, then split the data and feed part of it to each GPU in mini-batches. But as we know this means we need to copy and allocate memory for all of model's params, gradients and optimizer states on each GPU, which doesn't help with memory usage.

Reviewer 2: ```suggestion
1. **Data Parallelism** is what you get from `DataParallel` or `DistributedDataParallel` - we replicate the model to each GPU, then split the data and feed part of it to each GPU in mini-batches. But as we know this means we need to copy and allocate memory for all of model's parameters, gradients and optimizer states on each GPU, which doesn't help with memory usage.
```

> 2. And so ZeRO's ingenious approach is to give each GPU only a small chunk of params, gradients and optimizer states of each layer of the model. It doesn't assign groups of layers to GPUs, sometimes referred to as vertical slicing type of MP, but instead partitions (shards) each layer horizontally. This leads to zero overlap between GPUs. At runtime each GPU builds up each layer's data on the fly by asking participating GPUs to send the information it's lacking. This idea could be difficult to grasp, and you will find my attempt at explanation [here](https://github.com/huggingface/transformers/issues/8771#issuecomment-758418429).

Reviewer 3: ```suggestion
2. And so ZeRO's ingenious approach is to give each GPU only a small chunk of params, gradients and optimizer states of each layer of the model. It doesn't assign groups of layers to GPUs, sometimes referred to as vertical slicing type of MP, but instead partitions (shards) each layer horizontally. This leads to zero overlap between GPUs. At runtime each GPU builds up each layer's data on the fly by asking participating GPUs to send the information it's lacking. This idea could be difficult to grasp, and you will find my explanation attempt [here](https://github.com/huggingface/transformers/issues/8771#issuecomment-758418429).
```

Author: why doesn't what I wrote work? "attempt at something" is an idiom:
https://www.oxfordlearnersdictionaries.com/definition/english/attempt_

Is it too idiomatic and we want a simpler language?

Author: I think I should write it as "attempt at an explanation" though to be grammatically correct.

Or perhaps, let's scratch it altogether and write as simple as:

> If you find this idea difficult to grasp, you will find some explanations [here]

Reviewer 3: Thought "explanation attempt" sounds better, but just a nit

> 3. Finally, **Pipeline Parallelism** is added. As I explained [here](https://github.com/huggingface/transformers/issues/8771#issuecomment-758250421) a naive model parallelism suffers from a serious under-utilization of GPU resources, due to GPU idling, while waiting for other GPUs to finish their work. And PP solves this problem, but splitting mini-batches into micro-batches and feeding the idling GPUs new data while it's waiting for the other GPUs to complete their `forward` and `backward` stages.

Reviewer 2: ```suggestion
3. Finally, **Pipeline Parallelism** is added. As I explained [here](https://github.com/huggingface/transformers/issues/8771#issuecomment-758250421) a naive model parallelism suffers from a serious under-utilization of GPU resources, due to GPU idling, while waiting for other GPUs to finish their work. And PP solves this problem, by splitting mini-batches into micro-batches and feeding the idling GPUs new data while it's waiting for the other GPUs to complete their `forward` and `backward` stages.
```

> As of this writing FairScale and DeepSpeed only perform Model Parallelism (Sharding) for the optimizer states and gradients. Model params sharding is supposedly coming soon in DeepSpeed.

Reviewer 2: ```suggestion
As of this writing FairScale and DeepSpeed only perform Model Parallelism (Sharding) for the optimizer states and gradients. Model parameters sharding is supposedly coming soon in DeepSpeed.
```

The other powerful feature is ZeRO-Offload. This feature offloads some of the processing and memory needs to the host's CPU, thus allowing more to be fit onto the GPU. You saw its dramatic impact in the success at running `t5-3b` on a 24GB GPU.

One other problem that a lot of people complain about on pytorch forums is GPU memory fragmentation. One often gets an OOM error that goes like:
```
RuntimeError: CUDA out of memory. Tried to allocate 1.48 GiB (GPU 0; 23.65 GiB total capacity;
16.22 GiB already allocated; 111.12 MiB free; 22.52 GiB reserved in total by PyTorch)
```

> The program wants to allocate ~1.5GB and the GPU still has some 6-7GBs of unused memory, but it reports to have only ~100MB of contiguous free memory and it fails with the OOM (Out Of Memory) error. This happens as chunks of different size get allocated and de-allocated again and again, and over time holes get created leading to memory fragmentation, where there is a lot of unused memory but no contiguous chunks of the desired size. In the example above the program could probably allocate 100MB of contiguous memory, but clearly it can't get 1.5GB in a single chunk.

Reviewer 2: ```suggestion
The program wants to allocate ~1.5GB and the GPU still has some 6-7GBs of unused memory, but it reports to have only ~100MB of contiguous free memory and it fails with the OOM error. This happens as chunks of different size get allocated and de-allocated again and again, and over time holes get created leading to memory fragmentation, where there is a lot of unused memory but no contiguous chunks of the desired size. In the example above the program could probably allocate 100MB of contiguous memory, but clearly it can't get 1.5GB in a single chunk.
```
(was defined before)

> DeepSpeed attacks this problem by managing GPU memory by itself and ensuring that long term memory allocations don't mix with short-term ones and thus there is much less fragmentation. While the paper doesn't go into details, the source code is available, so it's possible to see how DeepSpeed accomplishes that.

Reviewer: I would add an additional link here when mentioning the source code:

```suggestion
DeepSpeed attacks this problem by managing GPU memory by itself and ensuring that long term memory allocations don't mix with short-term ones and thus there is much less fragmentation. While the paper doesn't go into details, the [source code](https://github.com/microsoft/DeepSpeed) is available, so it's possible to see how DeepSpeed accomplishes that.
```

As ZeRO stands for Zero Redundancy Optimizer, it's easy to see that it lives up to its name.

# The Future

Besides the anticipated upcoming support for model params sharding in DeepSpeed, it already released new features that we haven't explored yet. These include DeepSpeed Sparse Attention and 1-bit Adam, which are supposed to decrease memory usage and dramatically reduce inter-GPU communication overhead, which should lead to an even faster training and support even bigger models.

> I trust we are going to see new gifts from the FairScale team as well.

Author: I hope you don't mind if I keep "I trust". 

To me: "trusting" is having faith, "being sure" is something different, and here I can't be sure at all about them producing anything.

> I trust we are going to see new gifts from the FairScale team as well.

Reviewer 3: For sure all of those are just suggestions!

> I trust we are going to see new gifts from the FairScale team as well.

Reviewer 3: All nits :-)

# Issues

If you encounter any issues with the integration part of either of these projects please open an Issue in [transformers](https://github.com/huggingface/transformers/issues).

But if you have problems with DeepSpeed and FairScale installation, configuration and deployment - you need to ask the experts in their domains, therefore, please, use [DeepSpeed Issue](https://github.com/microsoft/DeepSpeed/issues) or [FairScale Issue](https://github.com/facebookresearch/fairscale/issues) instead.

# Resources

While you don't really need understand how any of these projects work and you can just deploy them via the `transformers` Trainer, should you want to figure out the whys and hows please refer to the following resources.

> * FairScale github https://github.com/facebookresearch/fairscale
> * DeepSpeed github https://github.com/microsoft/DeepSpeed

Reviewer 2: ```suggestion
* [FairScale GitHub](https://github.com/facebookresearch/fairscale)
* [DeepSpeed GitHub](https://github.com/microsoft/DeepSpeed)
```

> * Paper: ZeRO: Memory Optimizations Toward Training Trillion Parameter Models https://arxiv.org/abs/1910.02054. The paper is very interesting, but it's very terse.

Reviewer 2: ```suggestion
* Paper: [ZeRO: Memory Optimizations Toward Training Trillion Parameter Models](https://arxiv.org/abs/1910.02054). The paper is very interesting, but it's very terse.
```

* Here is a good [video discussion](https://www.youtube.com/watch?v=tC01FRB0M7w) of the paper with visuals

> * DeepSpeed configuration and tutorials: https://www.deepspeed.ai/getting-started/

Reviewer 2: ```suggestion
* DeepSpeed [configuration and tutorials](https://www.deepspeed.ai/getting-started/)
```

* In addition to the paper, I highly recommend to read the following detailed blog posts with diagrams:
  - [DeepSpeed: Extreme-scale model training for everyone]( https://www.microsoft.com/en-us/research/blog/deepspeed-extreme-scale-model-training-for-everyone/)
  - [ZeRO & DeepSpeed: New system optimizations enable training models with over 100 billion parameters](https://www.microsoft.com/en-us/research/blog/zero-deepspeed-new-system-optimizations-enable-training-models-with-over-100-billion-parameters/)
  - [Turing-NLG: A 17-billion-parameter language model by Microsoft](https://www.microsoft.com/en-us/research/blog/turing-nlg-a-17-billion-parameter-language-model-by-microsoft/)

> * DeepSpeed examples git https://github.com/microsoft/DeepSpeedExamples

Reviewer 2: ```suggestion
* DeepSpeed [examples on GitHub](https://github.com/microsoft/DeepSpeedExamples)
```

If you would like to experiment with this benchmark yourself or want to know more details about the hardware and software used to run it, please, refer to this [post](https://github.com/huggingface/transformers/issues/8771#issuecomment-759176685).

# The Magic Behind ZeRO

Since `transformers` only integrated these fabulous solutions and wasn't part of their invention I will share the resources where you can discover all the details for yourself. But here are a few quick insights that may help understand how ZeRO manages these amazing feats.

The key feature of ZeRO is adding distributed data storage to the quite familiar concept of data parallel training.

The computation on each GPU is exactly the same as data parallel training, but the parameter, gradients and optimizer states are stored in a distributed/partitioned fashion across all the GPUs and fetched only when needed.

The following diagram, coming from this [blog post](https://www.microsoft.com/en-us/research/blog/zero-deepspeed-new-system-optimizations-enable-training-models-with-over-100-billion-parameters/) illustrates how this works:

![ZeRO Partitioning](./assets/09_zero_deepspeed_fairscale/zero-partitioning.png]

> ZeRO's ingenious approach is to give each GPU only a small chunk of params, gradients and optimizer states of each layer of the model. It doesn't assign groups of layers to GPUs, sometimes referred to as vertical slicing type of model parallelism, but instead partitions (shards) each layer's data horizontally. This leads to zero overlap in data storage between GPUs. At runtime each GPU builds up each layer's data on the fly by asking participating GPUs to send the information it's lacking.

Reviewer 4: @stas00 this looks good to me. One minor comment I have is that the type of sharding (horizontal or vertical) does not really matter. ZeRO can work with either. Fairscale implements a vertical sharding, and DeepSpeed implements a combination of horizontal and vertical sharding.  May I suggest a mild rephrasing of what you already have to avoid the details of the sharding that might be confusing?  --> "ZeRO's ingenious approach is to partition the params, gradients and optimizer states equally across all GPUs and give each GPU just single partition.  This leads to zero overlap in data storage between GPUs. At runtime each GPU builds up each layer's data on the fly by asking participating GPUs to send the information it's lacking."

Author: That sounds good, @samyam. I was trying to emphases that it's not layer-partitioning, but your description is clear enough I think.

I only added to it a mentioning of a shard==partition, since the 2 are used interchangeably at times in some docs.

Thank you for your review and the suggested rewrite, @samyam

Reviewer 4: @stas00 we have been working on turning the ZeRO-Offload work into a technical paper, and it was uploaded to arxiv just today. For the blog readers who might be interested in learning more about the ZeRO-Offload technique, they might be able to find some additional details here: https://arxiv.org/abs/2101.06840 . Thought I would share in case you think its helpful to add to the resources for the blog.

Author: Amazing! Yes, of course - added it to the blog - just need to wait for those with commit rights to make it available on the website.

Thanks a lot for the heads up, @samyam!

Reviewer 4: Awesome! Thank you @stas00

Author: It's live!

Looking forward to reading the new paper.

You guys have created a lot of amazing tech, it'd be amazing if you were to hire someone to convey all the intricacies with lots of demos and simple tutorials to us users. It takes a lot of digging to understand things at the moment. And I'm concerned me bombarding you with many issues requesting to explain this and that is probably not easy to deal with.

Reviewer 4: Thank you @stas00. That is really good feedback. We will definitely look into it. Btw, you bombarding us with the many issues has really helped us improve, so please keep them coming.

Author: I appreciate the reassurance, @samyam.
